### PR TITLE
fix(intel): correctness fixes for CFG recovery (engine layout, stacked on #68)

### LINKED ISSUE #69: intel: getReferencedAddr drops sign of negative RIP-relative displacements
Surfaced by CodeRabbit on PR #68 (pre-existing in master; the engine refactor moved the helper verbatim).

`RecursiveDisassembler.getReferencedAddr()` matches `0x[a-fA-F0-9]+` and ignores a leading `-`, so an operand like `qword ptr [rip - 0x20]` is treated as `+0x20`. The x64 RIP-relative call/jmp paths in `X86Backend` build their slot address from this helper, so negative displacements resolve to the wrong slot and can mis-recover CFG/API edges.

Fix is behavior-changing (alters disassembly output), so it must land in its own PR with sha256 re-baselining of the integration fixtures, not in the zero-change refactor (#68). Suggested approach: capture an optional sign group and apply it.

### LINKED ISSUE #70: intel: RIP-relative qword call/jmp books the slot, not the in-image dereferenced target
Surfaced by CodeRabbit on PR #68 (pre-existing in master; moved verbatim into `X86Backend`).

For `call/jmp qword ptr [rip + X]`, both paths dereference the slot but keep CFG/tailcall bookkeeping on the slot address. When the qword points back into the image (thunk/local function), recursion never reaches the real target; only import/API cases should stay slot/API-only. The 32-bit `dword ptr [0x...]` call path already handles the in-image case. Behavior-changing → dedicated PR with re-baselining.

### LINKED ISSUE #71: intel: loop/loope/loopne taken edge is not enqueued
Surfaced by CodeRabbit on PR #68 (pre-existing in master; moved verbatim into `X86Backend._analyzeLoopInstruction`).

`loop*` are conditional branches with two successors, but only the fallthrough edge is queued; a forward `loop` target is never analyzed unless another path discovers it. Behavior-changing → dedicated PR with re-baselining.

### LINKED ISSUE #72: intel: minor robustness — address-0 truthiness checks and resolveIndirectSwitch None-guard
Surfaced by gemini-code-assist on PR #68 (all pre-existing in master; moved verbatim).

1. `if to_addr` / `if previous_address` truthiness checks skip the valid address `0` (only relevant for base-addr-0 buffers); explicit `is not None` is more correct.
2. `resolveIndirectSwitch` does `current_byte < size` where `current_byte` can be `None` if the offset leaves the image → `TypeError`. Note: this method currently has no callers (effectively dead) — consider removing or guarding.

Low priority; behavior-changing where reachable → batch into a follow-up.

### PR DESCRIPTION
**Pull request by r0ny123**
## Summary

Engine-layout port of the four CFG-recovery correctness fixes from upstream PR
danielplohmann/smda#141, applied to the new `RecursiveDisassembler` + `X86Backend` split.

**Stacked on #68** (base = `feat/arch-agnostic-engine`) so the refactor PR itself stays a pure,
zero-behavior-change move. Retarget to `master` once #68 merges. This branch is also exactly the
conflict resolution needed when upstream #140 is rebased over upstream #141 — whichever merges
first.

One commit per tracked issue:

| Commit | Issue | Fix |
|---|---|---|
| 1 | Fixes #69 | `getReferencedAddr()` preserves the displacement sign (`[rip - 0x20]` no longer parsed as `+0x20`); `JumpTableAnalyzer._x64Handler` drops its manual sign compensation |
| 2 | Fixes #70 | `call qword ptr [rip + X]` through a slot holding an **in-image** target now books the call on the dereferenced destination (mirrors the 32-bit `dword ptr [0x...]` path); import-like slots unchanged. The jmp path intentionally stays slot-based — same as the 32-bit jmp path, and redirecting it would interact with tailcall analysis |
| 3 | Fixes #71 | `loop`/`loope`/`loopne` taken edge is enqueued (conditional branches have two successors) |
| 4 | Fixes #72 | `is not None` guards for address-0 (`previous_address`, `to_addr`) and a `None` guard in `resolveIndirectSwitch` for the out-of-image walk (`TypeError`) |

## Validation

- 6 new tests in `tests/testIntelDisassembler.py`, each verified to fail before its fix and pass
  after (run through the `IntelDisassembler` wrapper, so they exercise the engine + backend path).
- `make lint` clean; `make test`: **117 passed**, including the byte-exact SHA-256 integration
  baselines — the fixture samples do not exercise these paths, so no re-baselining was needed.
- `capstone-semantics-reviewer` pass on the upstream twin: all four items verified OK against the
  pinned capstone (op_str formats incl. `bnd` prefix, `[rip]` zero displacement, negative-immediate
  formatting, `i_size` covering prefixes; `jcxz`-family confirmed in `CJMP_INS`, unaffected).

Closes #69, closes #70, closes #71, closes #72.

## Bug Fixes
- Fixed signed displacement parsing in address extraction for more accurate immediate operand handling
- Improved indirect switch scanning safety at memory boundaries
- Corrected address range validation to properly handle zero addresses
- Enhanced RIP-relative displacement calculation accuracy
- Improved indirect call target resolution
- Loop instructions now properly queue all branch destinations

## Tests
- Added comprehensive test coverage for disassembly edge cases and address parsing scenarios

### CODE REVIEW

File: src/smda/intel/X86Backend.py [hunk 1]
```diff
@@ -115,6 +123,8 @@ def _analyzeLoopInstruction(self, d, i, state):
         i_address, i_size, i_mnemonic, i_op_str = i
         jump_destination = d.getReferencedAddr(i_op_str)
         if jump_destination:
+            # loops are conditional branches: queue the taken edge as well
+            state.addBlockToQueue(jump_destination)
             state.addCodeRef(i_address, int(i_op_str, 16), by_jump=True)
```
[the remark this answers is not in the document]
Comment (r0ny123) [reply src/smda/intel/X86Backend.py:128]: Deliberately out of scope — this is the residual truthiness class already documented on #72 (see the closing comment there). Three reasons it stays:

1. `getReferencedAddr()` returns `0` both for "no hex literal found" *and* a genuine address 0 — the truthiness check is the pre-existing sentinel contract, shared verbatim with `_analyzeCondJmpInstruction` (line 108). Changing it only here would make the two conditional-branch handlers inconsistent; changing the contract properly means a None-sentinel across all callers, which is its own change.
2. The suggested `if "0x" in i_op_str:` doesn't actually fix the address-0 case: capstone prints address 0 as `0` (no `0x` prefix), so the branch would still be skipped — and `int(i_op_str, 16)` would be reached with no behavior difference vs `jump_destination`.
3. A `loop` whose taken target is exactly the image base of a base-0 buffer is pathological; this PR scopes #72 to the cases that silently disabled real detections (`previous_address`, `to_addr`).

Tracked as residual on #72. Resolving.

**Comment by r0ny123** [comment]
@coderabbitai full review
